lib/command.js

```diff
@@ -1257,81 +1449,93 @@ Expecting one of '${allowedValues.join("', '")}'`);
+
     operands = operands.concat(parsed.operands);
     unknown = parsed.unknown;
     this.args = operands.concat(unknown);
 
     if (operands && this._findCommand(operands[0])) {
-      return this._dispatchSubcommand(operands[0], operands.slice(1), unknown);
-    }
-    if (this._hasImplicitHelpCommand() && operands[0] === this._helpCommandName) {
-      return this._dispatchHelpCommand(operands[1]);
-    }
-    if (this._defaultCommandName) {
+      chain = this._chainOrCall(chain, () => this._dispatchSubcommand(
+        operands[0], operands.slice(1), unknown
+      ));
+    } else if (this._hasImplicitHelpCommand() && operands[0] === this._helpCommandName) {
+      chain = this._chainOrCall(chain, () => this._dispatchHelpCommand(operands[1]));
+    } else if (this._defaultCommandName) {
       outputHelpIfRequested(this, unknown); // Run the help for default command from parent rather than passing to default command
-      return this._dispatchSubcommand(this._defaultCommandName, operands, unknown);
-    }
-    if (this.commands.length && this.args.length === 0 && !this._actionHandler && !this._defaultCommandName) {
-      // probably missing subcommand and no handler, user needs help (and exit)
-      this.help({ error: true });
-    }
-
-    outputHelpIfRequested(this, parsed.unknown);
-    this._checkForMissingMandatoryOptions();
-    this._checkForConflictingOptions();
-
-    // We do not always call this check to avoid masking a "better" error, like unknown command.
-    const checkForUnknownOptions = () => {
-      if (parsed.unknown.length > 0) {
-        this.unknownOption(parsed.unknown[0]);
+      chain = this._chainOrCall(chain, () => this._dispatchSubcommand(
+        this._defaultCommandName, operands, unknown
+      ));
+    } else {
+      if (this.commands.length && this.args.length === 0 && !this._actionHandler && !this._defaultCommandName) {
+        // probably missing subcommand and no handler, user needs help (and exit)
+        this.help({ error: true });
       }
-    };
 
-    const commandEvent = `command:${this.name()}`;
-    if (this._actionHandler) {
-      checkForUnknownOptions();
-      this._processArguments();
+      outputHelpIfRequested(this, parsed.unknown);
+      this._checkForMissingMandatoryOptions();
+      this._checkForConflictingOptions();
 
-      let actionResult;
-      actionResult = this._chainOrCallHooks(actionResult, 'preAction');
-      actionResult = this._chainOrCall(actionResult, () => this._actionHandler(this.processedArgs));
-      if (this.parent) {
-        actionResult = this._chainOrCall(actionResult, () => {
+      // We do not always call this check to avoid masking a "better" error, like unknown command.
+      const checkForUnknownOptions = () => {
+        if (parsed.unknown.length > 0) {
+          this.unknownOption(parsed.unknown[0]);
+        }
+      };
+      const processAndAwaitArguments = () => {
+        checkForUnknownOptions();
```

Comment:
I didn't call `checkForUnknownOptions` here in my experiment in case not used in some branches, but in fact everywhere that `processAndAwaitArguments` is called, `checkForUnknownOptions` is the line before. So it is ok.

Reply:
Actually, `processAndAwaitArguments()` was originally simply called `preprocess()`, but then I decided the unknown option check and argument processing were too unrelated to keep them in one function, and also I noticed that `checkForUnknownOptions()` was called without `_processArguments()` in one place, so I wanted to remove the `checkForUnknownOptions()` call from the function, but forgot to do it and ended up only renaming the function.

So we should either remove the `checkForUnknownOptions()` call from `processAndAwaitArguments()` as planned, or rename the function back to `preprocess()` or something along these lines and remove the redundant `checkForUnknownOptions()` calls. What do you prefer?

Reply:
I prefer removing `checkForUnknownOptions` from `processAndAwaitArguments` too, not related enough so harder to understand what is being checked where.

(As understanding of the pattern evolves we might come up with a routine name where it makes sense included again!)

Reply:
Resolved in [905dac0](https://github.com/tj/commander.js/pull/1915/commits/905dac07a17d017029afc7d1a7de1540ed93a228).